TST: Add cygwin build to CI

DWesl opened the pull request:
Attempt to add a Cygwin build to the CI.  

The code builds just fine, but the actual "running the tests" bit is crashing for reasons I don't understand and can't reproduce on my local machine.  One way around this would be to build a wheel, install that into a virtual environment, and test using `np.test()`.  Another would be using `tox` to run the tests, basically automating the previous suggestion.

Inspired by discussion on #18308. 
<!--         ----------------------------------------------------------------
                MAKE SURE YOUR PR GETS THE ATTENTION IT DESERVES!
                ----------------------------------------------------------------

*  FORMAT IT RIGHT:
      http://www.numpy.org/devdocs/dev/development_workflow.html#writing-the-commit-message

*  IF IT'S A NEW FEATURE OR API CHANGE, TEST THE WATERS:
      http://www.numpy.org/devdocs/dev/development_workflow.html#get-the-mailing-list-s-opinion

*  HIT ALL THE GUIDELINES:
      https://numpy.org/devdocs/dev/index.html#guidelines

*  WHAT TO DO IF WE HAVEN'T GOTTEN BACK TO YOU:
      http://www.numpy.org/devdocs/dev/development_workflow.html#getting-your-pr-reviewed
-->

DWesl committed:
TST: Try to set up a test environment on Cygwin.

Edit: Fix spelling of "pull request"

TST: Follow workflow for other OSs using commands available on Cygwin.

Other GitHub actions set fetch-depth to zero and use ./.github/actions
for the actual testing.  Let's see if this works.

Edit: Drop shell from the last step in the action.

Apparently "shell" isn't valid for steps.  I hope the action knows to
use bash rather than Power Shell or cmd.exe, or this is going to get
very confused.

TST: Adjust for cygwin

The action assumes `sudo` exists, which is a bit of a problem.

TST: Fetch with Cygwin git

Make sure Cygwin git can access the repository.  Versioneer depends on
git, so this is kind of important.

Edit: Include closing quote.

Edit: Install Cygwin git.

Versioneer needs this in order to function.

Edit: FIXUP: build_src option is not spelled --verbose-config

It's --verbose-cfg.

TST: Check which version of Python is being used.

I want to be sure it's a Cygwin one, not a Windows one.

TST: Build a wheel for Cygwin.

I may need to install in a virtualenv, and installing from a wheel is
much faster than installing from a directory.

TST: Use runtests.py to build the extensions.

It couldn't find them during the tests, which is not a problem I've
run into locally.  Hopefully this fixes that.

TST: Drop separate build step.

TST: Cygwin CI: Install numpy and run tests from that.

Running with runtests.py didn't work, but this should.

TST: Add cygwin bin dir to path, not root dir.

TST: Make sure Cygwin build fails when tests fail

The tests managed to pass despite NumPy not importing.  This is not a good thing.

TST: Avoid steps after tests in Cygwin CI.

I really need this to fail if NumPy doesn't import.

TST: Ensure Cygwin CI is running in Cygwin.

Kinda defeats the point if it's not.

TST: Check that pip installed the C extensions.

This really shouldn't need to be checked, but the CI runs have been
failing due to a failed import of a C extension.  I also printed the
version, which will make me feel better about the right version being
found and used.

TST: Include closing quote in test command

TST: Report loaded modules

Still trying to find why the C extension module import fails.

TST: Check permissions on NumPy C extensions

TST: Ask python which files it's trying to load.

Hopefully this gives me a general direction for where to look for
"Cannot find file or directory"

TST: Work around powershell syntax weirdness

Drop a level of quoting and just write the command to a file, then run that.  It should still show DLL permissions.

TST: Fix line endings in script.

bash expects \n
powershell creates \r\n

TST: Check for DLLs required by C extensions in Cygwin CI.

It's still not working; hopefully this shows why.

TST: Add import checks to the dll testing.

I have a project where `python -c 'import module'` worked but
`pytest --pyargs module` did not.  Lets see if that happens here.

Allow for global installs

The runners have global install privileges, so pip will install NumPy there.  I need to account for that in my script.

Fix sed regex.

Lots of leaning-toothpick syndrome, but it actually does what it's supposed to locally.  Hopefully the CI agrees.

Stop trying to import NumPy modules from sourcedir

This doesn't work and hasn't for a while.

TST: Simplify PATH and make sure lapack is on it.

Most recent CI run said it couldn't find lapack.
Reduce PATH to just Cygwin directories, make sure /usr/lib/lapack is included (after /usr/bin), and try again.

Shortening PATH occasionally fixes some other problems.  Let's see if that works here.

TST: Change Cygwin CI python from 3.8 to 3.7

Let's see if this solves the problems.  There was a change in DLL load path handling in 3.8 that might be causing the "cannot load numpy.linalg.lapack_lite" errors.

TST: Stop running CI on PR branch.

STY: Wrap long lines in Cygwin workflow file.

TST: Update name of "main" branch used to trigger workflow.

I forgot this changed a while back.

TST: Specify full paths for commands.

Also use dash in more places.

TST: Change the newline-escape mechanism in Cygwin workflow.

Backslash is apparently difficult.  This reduces that step to a single
command and tells YAML the string is to be interpreted as a single
line.

TST: Move the test command onto one line so GitHub can read it.

Apparently the GitHub Actions YAML parser is incomplete.  It should be
reading what was there as a single-line string with no linebreak at
the end and all linebreaks in the middle replaced by spaces.  It seems
to have parsed this as two lines.

Does GitHub have a document saying which subset of YAML they actually
recognize?

TST: Add dependency on importlib-metadata.

Pytest didn't declare it.

TST: Move importlib-metadata earlier in install list.

I hope this means it actually gets installed.  The manager is ignoring
it right now.

TST: Add zipp to the package install list.

Also tell pip to install the test_requirements, so I don't keep
running into this problem.

TST: Specify full path to python for testing.

I apparently missed this earlier.

TST: Add CFFI and pytz to Cygwin test environment.

TST: Set global path to include /usr/lib/lapack.

I'd forgotten there was a line for this already.

TST: Shorten and unpin the Cygwin test requirements.

I don't want to build more modules than I have to.  It works fine on
my machine with the system setuptools, pytz, and CFFI as well as
system-ish Cython, so it should work fine on the CI runners.

TST: Make sure test requirements are actually installed.

Apparently
pytest->importlib_metadata->"typing_extensions; python_version <= 3.7"
isn't a declared dependency chain.  Mentioning the last two
explicitly should get them installed anyway.

TST: Fix quoting in command line.

DWesl commented:
I suppose I should mention: `python runtests.py` gets through the build process fine, and I'm assuming it starts the testing process because it breaks when trying to import `numpy.linalg` because it can't find `lapack_lite` or `_umath_linalg`.  I'm assuming the other CI has a way of setting things up so those modules are found, but I'm not sure what that is.

I've been using `python runtests.py` from the NumPy root directory for testing on Cygwin for a while now without issues at this point.  I'm not sure what's going on here, and have not been able to reproduce this locally.

mattip commented:
I wonder what is up with the doc build?
```
rm -rf build/*
find . -name generated -type d -prune -exec rm -rf "{}" ";"
# for testing
# @echo installed numpy 4d148aa1a9 matches git version 4d148aa1a9; exit 1
mkdir -p build
touch build/generate-stamp
mkdir -p build/html build/doctrees
LANG=C sphinx-build -b html -WT --keep-going -d build/doctrees  -q source build/html 
1.21.dev0 1.21.0.dev0+677.g4d148aa1a
make: *** [Makefile:179: html-build] Terminated

Too long with no output (exceeded 10m0s): context deadline exceeded
```

DWesl commented:
> I wonder what is up with the doc build?
> 
> ```bash
> LANG=C sphinx-build -b html -WT --keep-going -d build/doctrees  -q source build/html 
> 1.21.dev0 1.21.0.dev0+677.g4d148aa1a
> make: *** [Makefile:179: html-build] Terminated
> 
> Too long with no output (exceeded 10m0s): context deadline exceeded
> ```

No idea.  I'm hoping it's a transient thing that goes away with the new commit.

I tried building a NumPy wheel, installing it, and testing against that, but is still failed to import `numpy`.  I think the new version will get the run status to reflect that.  I'm still not sure why this isn't working, since those are the steps used to generate the NumPy I'm currently using on Cygwin.  I'm running rather low on ideas at this point.

seberg commented:
It has been going on a lot, I have a PR open to fix it (sorry missed the discussion here).  My theory is that we are getting different hardware so that it now takes sometimes up to 30 instead of around 10 minutes, and so we suddenly get timeouts frequently.  In any case, seems that bumping the limit in gh-18349 should make it reliable again.

DWesl commented:
The test is failing with "ImportError: No such file or directory" after attempting to import numpy.linalg.lapack_lite.  A previous step said the file it's trying to open (`/usr/local/lib/python3.8/site-packages/numpy/linalg/lapack_lite.cpython-38-x86_64-cygwin.dll`) exists and has the right permissions (read and execute for everyone).  

I could try a user install, a virtual environment, or tox, but I'm not sure those would make a difference.  I'm still not sure what's going on here.

mattip commented:
Unfortunately, windows does not have tooling like Linux to detect which DLLs are missing when loading a DLL at runtime. Does cygwin provide an `ldd`-like command to do anything like that? It might be something around the fortran or openblas dll.

DWesl commented:
> Unfortunately, windows does not have tooling like Linux to detect which 
> DLLs are missing when loading a DLL at runtime. Does cygwin provide an 
> |ldd|-like command to do anything like that? It might be something 
> around the fortran or openblas dll.
> 

It provides |ldd| and an alternate command I'm more familiar with.  I 
added both to the check on the dlls.

DWesl commented:
`ldd` reports no missing dependencies for any of the NumPy-installed DLLs as reported by pip.  All of the DLLs also have execute permissions for all users, and report no attributes that would change the permissions from 0755.  

Is the failure to load something specific to Windows or Docker or the combination?

charris commented:
> Is the failure to load something specific to Docker or Docker running Windows?

I don't know, but Windows has been tricky in that regard in the past. What we need here is some input from a Windows person.

DWesl commented:
> > Is the failure to load something specific to Docker or Docker running Windows?
> 
> I don't know, but Windows has been tricky in that regard in the past. What we need here is some input from a Windows person.

Is there some flag we need to set to tell Docker or Windows that we want to use this file as a DLL, despite it being created in this container?

charris commented:
Python 3.8 changed the way load paths were specified on Windows, so that might be related.  See ([permalink](http://numpy-discussion.10968.n7.nabble.com/Compile-NumPy-with-ifort-MSVC-and-MKL-DLL-load-failed-tp48968p48970.html)). @cgohlke Thoughts?

DWesl commented:
## A bit of background on Cygwin BLAS and LAPACK
The netlib reference BLAS and LAPACK dlls are installed to `/usr/lib/lapack`.  The OpenBLAS dll is installed to `/usr/bin` under the same name as the reference BLAS dll.  The default setup puts `/usr/bin` before `/usr/lib/lapack` in the `PATH`, so Windows finds OpenBLAS and loads that instead of the reference BLAS.

## Attempts to load BLAS and LAPACK from Python3.8 on my local Cygwin
```bash
$ ls /usr/bin/cygwin1.dll
/usr/bin/cygwin1.dll

$ python3.8
Python 3.8.7 (default, Jan 31 2021, 21:50:45)
[GCC 10.2.0] on cygwin
Type "help", "copyright", "credits" or "license" for more information.
```
```python
>>> import os
>>> os.add_dll_directory
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: module 'os' has no attribute 'add_dll_directory'
>>> import ctypes.util
>>> ctypes.CDLL("cygwin1.dll").cygwin_conv_path
<_FuncPtr object at 0x6fffffe6eac0>
>>> ctypes.CDLL(ctypes.util.find_library("lapack")).dgesv_
<_FuncPtr object at 0x6ffff790dd00>
>>> import numpy.linalg.lapack_lite
>>>
```
The lack of `os.add_dll_directory` on  Cygwin implies the DLL load path didn't change.
If `/usr/lib/lapack` not being on the rewritten DLL load path list were an issue, one of the LAPACK loads would fail.  The load issue might be that `/usr/lib/lapack` isn't on the PATH, but that would probably show up in the `ldd` output.  

I could try switching to 3.7 if you think that would help.

charris commented:
> I could try switching to 3.7 if you think that would help.

It would eliminate one possible concern. Have you tried searching stackoverflow?

DWesl commented:
I tried with python 3.7:
```
ls
-rwxr-xr-x 1 runneradmin None   194903 Feb 18 18:52 /usr/lib/python3.7/site-packages/numpy/linalg/lapack_lite.cpython-37m-x86_64-cygwin.dll
file
/usr/lib/python3.7/site-packages/numpy/linalg/lapack_lite.cpython-37m-x86_64-cygwin.dll:       PE32+ executable (DLL) (console) x86-64, for MS Windows
ldd
/usr/lib/python3.7/site-packages/numpy/linalg/lapack_lite.cpython-37m-x86_64-cygwin.dll:
	ntdll.dll => /cygdrive/c/windows/SYSTEM32/ntdll.dll (0x7ffd59240000)
	KERNEL32.DLL => /cygdrive/c/windows/System32/KERNEL32.DLL (0x7ffd574e0000)
	KERNELBASE.dll => /cygdrive/c/windows/System32/KERNELBASE.dll (0x7ffd560b0000)
	lapack_lite.cpython-37m-x86_64-cygwin.dll => /usr/lib/python3.7/site-packages/numpy/linalg/lapack_lite.cpython-37m-x86_64-cygwin.dll (0x5cde30000)
	libpython3.7m.dll => /usr/bin/libpython3.7m.dll (0x3fc150000)
	cygwin1.dll => /usr/bin/cygwin1.dll (0x180040000)
cygcheck
C:\tools\cygwin\lib\python3.7\site-packages\numpy\linalg\lapack_lite.cpython-37m-x86_64-cygwin.dll
  C:\tools\cygwin\bin\cygwin1.dll
    C:\windows\system32\KERNEL32.dll
      C:\Program Files\PowerShell\7\api-ms-win-core-rtlsupport-l1-1-0.dll
      C:\windows\system32\ntdll.dll
      C:\windows\system32\KERNELBASE.dll
        C:\Program Files\PowerShell\7\api-ms-win-eventing-provider-l1-1-0.dll
      C:\Program Files\PowerShell\7\api-ms-win-core-processthreads-l1-1-0.dll
      ...
      C:\Program Files\PowerShell\7\api-ms-win-core-psapi-ansi-l1-1-0.dll
  C:\tools\cygwin\bin\libpython3.7m.dll
    C:\tools\cygwin\bin\cygintl-8.dll
      C:\tools\cygwin\bin\cygiconv-2.dll
    C:\tools\cygwin\bin\cyggcc_s-seh-1.dll

python -vv
# trying /usr/lib/python3.7/site-packages/numpy/linalg/lapack_lite.cpython-37m-x86_64-cygwin.dll
Traceback (most recent call last):
    ...
ImportError: No such file or directory
```
Same error.  None of my searches ("GitHub Actions on Windows", "Problems loading DLLs in GitHub Actions on Windows", or "GitHub Actions on Windows third-party DLL") returned anything relevant.

mattip commented:
Could you save the wheel as a [github action artifacts](https://github.com/actions/upload-artifact)and then download that to a local machine to test it? Maybe something is missing in the packaging

DWesl commented:
I installed a snapshot of the Cygwin dll a few months back to get around a segfault in `modfl`.  This means `uname -r` starts with `3.2.0` rather than the `3.1.7` of the current release.  As a consequence, pip doesn't recognize the uploaded wheel as valid for my platform (it also doesn't recognize wheels built with Cygwin 3.1.6 as valid on 3.1.7, but that's not relevant here).

Does anyone know how to get pip to ignore this and install the wheel anyway?  I got `numpy.linalg.lapack_lite` to import by unpacking with `wheel` then adding execute permissions to all installed dlls.  `np.linalg.test()` passes, so loading `lapack_lite` seems to work, narrowing the problem to some difference between my system and the GitHub runners.

DWesl commented:
I seem to have a similar problem with a different project on my computer.  I still have no idea why it's not importing the file it finds with the proper permissions, but I can investigate this locally now.

charris commented:
Great!

DWesl commented:
Thanks @mattip for suggesting I look at `ldd` output.  I eventually saw problems finding lapack when looking at the extension modules, so I explicitly set `PATH=/bin:/usr/bin:/usr/lib/lapack` when looking at the extension modules and running the tests.
It's still failing, but it's starting to run the tests first, so I think it's just regular failure now (as opposed to CI failures not reproducible locally).  Unfortunately, the failure is a new segfault I need to track down.

Does anyone want me to rebase and squash some of the comments together or make the commit messages numpy-style (add TST: at start)?  I've recently had a couple rounds of "try to do something, then account for differences between local and CI systems"

charris commented:
> Does anyone want me to rebase and squash some of the comments [commits?]

That would be good and you might as well rewrite the commit message(s) while doing so. The tests are passing now, do you think this is about ready?

charris commented:
@mattip @seberg Ready for review.

charris commented on .github/workflows/cygwin.yml:
```diff
@@ -0,0 +1,73 @@
+          bash -c "PATH=/bin:/usr/bin:/usr/lib/lapack bash list_dlls_unix.sh"
+      - name: Run NumPy test suite
+        run: |
+          dash -c "mkdir empty"
```
Is `dash` correct?

DWesl replied:
[Dash](http://gondor.apana.org.au/~herbert/dash/) is an alternate implementation of the Bourne shell.  It is faster than bash on Cygwin for many applications.  The website says it tries to be small and fast, and being small will help make `fork()` faster on Cygwin.  

I can change over to bash if you prefer, or let PowerShell execute that line since there's no files being written in that directory.

charris replied:
No problem, I just thought it might be a typo. I've used dash during failed linux boots,

DWesl replied:
I added `dash` to the requirements to make sure it's available here.

charris commented on .github/workflows/cygwin.yml:
```diff
@@ -0,0 +1,73 @@
+          for name in ${dll_list};
+          do
+              echo ${name}
+              python3.7 -c "import "$(echo ${name} | sed -E -e "s/\/+(home|usr).*?site-packages\/+//g" -e "s/\//./g" -e "s/.cpython-3.m?-x86(_64)?-cygwin.dll//g")
```
Is it possible to break this line (and similar) with `\`?

DWesl replied:
Some places, yes, other places, not easily.

I could try using `runtests.py` again; that should work now and would be shorter than the current `np.test()` line

charris commented on .github/workflows/cygwin.yml:
```diff
@@ -0,0 +1,73 @@
+        with:
+          platform: x64
+          install-dir: 'C:\tools\cygwin'
+          packages: python37-devel python37-pytest python37-hypothesis python37-cython python37-pip python37-wheel python37-setuptools liblapack-devel libopenblas gcc-fortran git
```
It should be possible to break this line with `packages: >`. See https://stackoverflow.com/questions/3790454/how-do-i-break-a-string-in-yaml-over-multiple-lines.

DWesl replied:
Done

charris commented on .github/workflows/cygwin.yml:
```diff
@@ -0,0 +1,82 @@
+                  -e "s/\/+(home|usr).*?site-packages\/+//g" -e "s/\//./g" \
+                  -e "s/.cpython-3.m?-x86(_64)?-cygwin.dll//g")
+          done
+          '
```
Stray?

charris replied:
I was wondering about `'`.

charris replied:
But now I see it starts way up above :)

DWesl replied:
Yes, that's the end of the extension-module-checking script.  Should I split that into a separate file somewhere rather than writing it out each time?

charris replied:
I think the first thing is to get the test passing again, we can worry about further cleanups after that. I'm beginning to regret the original suggestion :) But it does look _much_ nicer now.

DWesl committed:
TST: Clean up the output from the extension module check.

Also serves to document the script a bit.

DWesl commented:
This is now failing within the actual NumPy test suite, somewhere inside `numpy/core/tests/test_simd.py`.  

If I use `runtests.py` rather than `np.test()` to launch the tests, I can split the tests in that file from the rest of the test suite.  The majority of the test suite has no segfaults and can run quickly using the default runner, while the simd tests do have segfaults and need to be run with pytest-forked to provide details on which tests are failing, especially if the tests are parameterized.

Should I switch to using `runtests.py` again?  Should I split out the tests in `numpy.core.tests.test_simd` from the rest of the suite so we actually get an error count for most of the code base?

charris commented:
I don't know what you should do. Were the actual tests running before? Runtests is probably the best interface for debugging, just use `-n` (test installed wheel) and `-vv` (more verbosity).

charris commented:
Might be nice to know what compiler version Cygwin is using.

DWesl commented:
The compiler should be GCC 10.2.0. 

 I forgot to change the main branch name from `master` to `main`, so there are a decent number of commits where the workflow wasn't run on the PR.  There are still the test failures that have been around for the last few NumPy versions my other PR is aimed at fixing, plus the segfaults in test_simd that I didn't see in older versions.

charris commented:
The segfault is in ../../core/tests/test_simd.py::Test_SIMD_BOOL_256_FMA3__AVX2_b8::test_operators_logical. Maybe @seiko2plus can help.

seiko2plus commented:
@charris, I will dig into it

seiko2plus commented:
> ../../core/tests/test_simd.py::Test_SIMD_BOOL_256_FMA3__AVX2_b8::test_operators_logical

```Bash

│   0x44e2a2e29 <PySIMDVector_AsData+41>    callq  0x44e3a39e8 <PyObject_IsInstance>                                                                                                                             │
│   0x44e2a2e2e <PySIMDVector_AsData+46>    test   %eax,%eax                                                                                                                                                     │
│   0x44e2a2e30 <PySIMDVector_AsData+48>    je     0x44e2a2ea0 <PySIMDVector_AsData+160>                                                                                                                         │
│   0x44e2a2e32 <PySIMDVector_AsData+50>    mov    0x10(%rsi),%eax                                                                                                                                               │
│   0x44e2a2e35 <PySIMDVector_AsData+53>    cmp    %ebx,%eax                                                                                                                                                     │
│   0x44e2a2e37 <PySIMDVector_AsData+55>    jne    0x44e2a2e60 <PySIMDVector_AsData+96>                                                                                                                          │
│   0x44e2a2e39 <PySIMDVector_AsData+57>    vmovdqu 0x20(%rsi),%xmm1                                                                                                                                             │
│   0x44e2a2e3e <PySIMDVector_AsData+62>    vinserti128 $0x1,0x30(%rsi),%ymm1,%ymm0                                                                                                                              │
│  >0x44e2a2e45 <PySIMDVector_AsData+69>    vmovdqa %ymm0,(%r12)  
```
GCC tries to aligned load local variable of 256-bit vector from non-aligned 256-bit stack pointer.
The bug has been reported multiple times and only limited for GCC on windows, see:
https://gcc.gnu.org/bugzilla/show_bug.cgi?id=49001
https://gcc.gnu.org/bugzilla/show_bug.cgi?id=40838
https://gcc.gnu.org/bugzilla/show_bug.cgi?id=54412

It seems this bug will never be fixed and the only workaround is to avoid dealing with GCC ABI when
it comes to data types that requires alignment larger than 128-bit.
In other words force inlining any function that accepts or returning data types of __m256/__m512, or just pass or returns any variables by constant reference.
Another aggressive solution would be force disable AVX/AVX512 on GCC/Windows.

charris commented:
@seiko2plus Thanks for the update. Sounds like disabling that for gcc on windows would be the way to go. We do produce 32 bit windows wheels using mingw, but I suppose avx is not a problem on that platform.  @mattip Do you know if we actually use gcc to produce windows wheels? I'm pretty sure we use MSVC for 64 bits.

seiko2plus commented:
@charris, Disabling AVX may lead to a significant loss of performance, I would prefer to work under this BUG instead. all we need is to avoid exporting any functions that accept or return AVX registers, SIMD interface already force inlining all intrinsics. #18795 should pass the SIMD test.

charris commented:
@seiko2plus  Sounds good.

charris commented:
Close/reopen

charris commented:
Different segfault in `../../core/tests/test_multiarray.py::TestArgmax::test_all `

charris commented:
Let's try again.

DWesl commented:
```
Fatal Python error: Segmentation fault

Current thread 0x0000000800000010 (most recent call first):
  File "/usr/lib/python3.7/site-packages/numpy/core/_methods.py", line 40 in _amax
  File "/usr/lib/python3.7/site-packages/numpy/core/tests/test_multiarray.py", line 4265 in test_all
```
<!--
https://github.com/numpy/numpy/blob/a1b8d3c1d54d9e6efb79492754f0744be4c8a353/numpy/core/tests/test_multiarray.py#L4262-L4269
https://github.com/numpy/numpy/blob/a1b8d3c1d54d9e6efb79492754f0744be4c8a353/numpy/core/_methods.py#L38-L40
https://github.com/numpy/numpy/blob/a1b8d3c1d54d9e6efb79492754f0744be4c8a353/numpy/core/_methods.py#L18
https://github.com/numpy/numpy/blob/a1b8d3c1d54d9e6efb79492754f0744be4c8a353/numpy/core/_methods.py#L10
https://github.com/numpy/numpy/blob/a1b8d3c1d54d9e6efb79492754f0744be4c8a353/numpy/core/umath.py#L10
https://github.com/numpy/numpy/blob/a1b8d3c1d54d9e6efb79492754f0744be4c8a353/numpy/core/setup.py#L960-L974
https://github.com/numpy/numpy/blob/a1b8d3c1d54d9e6efb79492754f0744be4c8a353/numpy/core/setup.py#L960-L974
-->
I don't know NumPy internals well enough to trace that back to its source unless the problem is just that this function needs  `CYG_FINLINE`:
https://github.com/numpy/numpy/blob/914407d51b878bf7bf34dbd8dd72cc2dbc428673/numpy/core/src/umath/loops.h.src#L215-L225

DWesl commented:
For more fun, running the test suite for NumPy `main` completes with only the expected test failures (two new ones since 1.16 or so) and no segfaults, so this seems to be something peculiar to the runner.

charris commented:
Yay, the segfaults are gone. Thanks Sayed. @DWesl Now to figure out why the tests are failing. I notice that a bunch of them are with complex256, not sure why. Does Cygwin actually have float128 on windows? You might try block listing  some Cygwin functions in `numpy/core/src/common/npy_config.h`.

DWesl commented:
Cygwin has float128 and complex256, but I think the C math functions got pulled from somewhat buggy versions in Mingw-w64 a while back, and haven't been updated as the upstream versions got fixed unless someone points out that would be useful.  About half those failures are related to `hypotl(inf, nan) = cabsl(inf + nan*j) = nan`, in either order and with either sign for the infinity, and one of the others is `hypotl(LDBL_MAX, 0.0L) = inf`.

I summarized the failures a few months ago at the start of #18308: should I cherry-pick the relevant commits over to this PR? I think cfa9b4e380d6cd396455f63a57891bec8d4bfe48 does npy_config.h stuff, though I could go through and check which functions are still relevant.

seiko2plus commented:
@charris,
> Different segfault in ../../core/tests/test_multiarray.py::TestArgmax::test_all

segfault caused due to the same GCC/WIN64 bug but this time with `AVX512`, here the error report from Intel SDE

```Bash
>>> a = np.random.normal(0, 1, (4, 5, 6, 7, 8))
>>> for i in range(a.ndim):
...             amax = a.max(i)

SDE ERROR:  TID: 0 executed instruction with an unaligned memory reference to address 0xffffb5d0 INSTR: 0x46e8d7706: IFORM: VMOVAPD_MEMf64_MASKmskw_ZMMf64_AVX512 :: vmovapd zmmword ptr [rax], zmm0
        IMAGE:    \\?\C:\cygwin64\lib\python3.7\site-packages\numpy\core\_multiarray_umath.cpython-37m-x86_64-cygwin.dll
        FUNCTION: .text
        FUNCTION ADDR: 0x46e8b8a60
```
@DWesl,
> (two new ones since 1.16 or so) and no segfaults, so this seems to be something peculiar to the runner.

That because segfaults will only appears on X86 CPUs with AVX512 support


@charris,
> Yay, the segfaults are gone. Thanks Sayed

My pleasure, the segfaults should gone but I'm not sure yet since the current CI/Worker doesn't support `AVX512`.

DWesl committed:
BLD: Tell NumPy about functions that cause test failures.

Several of the test failures still happen.  I'm not entirely sure why.

BLD: Add more functions to the Cygwin replace list.

These functions are mentioned in test failures, so I mark them for
replacement, along with the more obvious functions that might get
called by functions that continue to fail.

TST: List more functions to be replaced on Cygwin.

casin and casinf don't pass the branch cut tests.  Let's see if
replacing them lets the test pass on CI.

TST: Mark more functions to be replaced on Cygwin.

I tried to note the tests each function fails by group, but I don't
remember all of them anymore.

Let's see if casin{,f,l} gets replaced on the CI runner.

BLD: Mark more functions for replacement on Cygwin.

I probably need to run `git clean` to see improvements locally.

BLD:List more functions to be replaced on Cygwin.

This is nearly the last of them.

There are still a few failures I don't know how to deal with.
The cabsl/hypotl overflows will be gone next Cygwin update.
I may have an idea for cpowl (if cpowl(x, n) == cexpl(n*clogl(x)))
I don't understand timezones, CFFI, or LAPACK,
and that's the rest of the failures.

BLD: Change list of functions marked for replacement on Cygwin.

I hoped this would convince `cpowl` to flag its overflows, but that
doesn't appear to have happened.

charris commented:
close/reopen

DWesl commented:
I told NumPy about the functions mentioned in the test failures, which got rid of some of the failures.  Blocking `hypotl` as well as `cabsl` might eliminate most of the `np.complex256` failures.  I'm not sure why blocking `casinh` and friends leaves the failures in `test_loss_of_precision`, especially since `casin` seems like it wouldn't fail this way.  

I'm not sure what direction this PR should go from here, since I don't understand why the remaining failures are still there.

charris commented:
close/reopen

DWesl commented:
So, fun thing happened when trying to add 
```C
#undef HAVE_CABSL
#undef HAVE_HYPOTL
```
to the current version: with the above code in `npy_config.h`, I get segfaults GDB says are in `PySIMDVector_AsData` called from `simd_arg_converter`.  Without them (that is, with the current version on GitHub), I get segfaults GDB says are in `simd_arg_to_obj` called from `simd.intrin_setall_u8`.  GDB says there's no symbol table available, so I'm not sure how accurate those are.  How would I go about debugging this further?

This machine has AVX2 but not AVX512, from what I can tell.

charris commented:
That's interesting. The fallback cabsl calls hypotl, which falls back to double precision and calls sqrt, which is where I think SIMD creeps in. How does Cygwin treat long double? Windows defines it as double, I wonder if that is an option for Cygwin?

DWesl commented:
I should probably mention: pytest says the crash is in `numpy.core.tests.test_simd.test_operators_logical`.  I just pushed the commit marking `cabsl` and `hypotl` for replacement, which should take the number of CI test failures down a bit.  If I run
```bash
PATH=/bin:/usr/bin:/usr/lib/lapack NPY_BLAS_ORDER=blas NPY_LAPACK_ORDER=lapack python runtests.py -- --ignore=numpy/core/tests/test_simd.py
```
the test suite completes without segfaults and only nine failures.  
- `cpowl` doesn't report overflow, even when `cpowl` and `powl` are marked for replacement.
- `cabsl` reports overflow in some cases it should complete (with `cabsl` and `hypotl` marked for replacement)
- `csqrt` and `csqrtf` don't do branch cuts properly when marked for replacement
- `casinhf`, `casinh`, and `casinhl` lose precision for small arguments
- `SVD` doesn't behave like it should
- CFFI can't find `numpy.random._generator.random_standard_normal_fill`

Cygwin uses the Newlib implementations for `long double` functions by default.  Newlib seems targeted at systems without much memory, so the `long double` implementations there tend to be recent and simple.  On some occasions, people have made the case that Cygwin should have better implementations, at which point the maintainers often grab the ones from the Mingw-w64 project.  

I have no idea how one would go about disabling long double for a NumPy build, but it would likely get the CI passing if you want to do that.

DWesl commented:
It's down to four failures now:
- `numpy.core.tests.test_datetime.TestDateTime.test_datetime_as_string_timezone`: Expected `01:30-05`, Got `06:30+00`
- `numpy.core.tests.test_scalarmath.TestAbs.test_builtin_abs[complex256]`: `abs(HUGE+0j) == inf`, not HUGE
- `numpy.core.tests.test_scalarmath.TestAbs.test_numpy_abs[complex256]`: Encountered overflow in absolute
- `numpy.linalg.tests.test_linalg.TestCond.test_nan`: computing SVD of a matrix with a `nan` fails to converge

I don't know why these two `cabsl` tests still fail when all the other ones pass, and I don't know enough about the other two to know why those are failing.  Any ideas?

I've made sure pytz is installed on the runner.  Is NumPy expecting another timezone package to be available as well?

DWesl committed:
TST: Return to requirements in test_requirements.txt

There will probably be compilation involved for coverage (from
pytest-cov), maybe also cython, but the rest look like pure python and
should be fast.

DWesl committed:
BLD: Export random distribution functions on Cygwin.

CFFI should be able to find them now.  The Cygwin runtime DLL loader
is the Windows one, and the linker also shares most of the same
semantics.

DWesl commented:
Down to three errors now.  I can find the backup definition for `npy_cabsl` in terms of `npy_hypotl`: 
https://github.com/numpy/numpy/blob/0f7f313d847e8e36e9fb3c4e576619a3af3bfe56/numpy/core/src/npymath/npy_math_complex.c.src#L121-L127
but I can only find a definition for `npy_hypot`, not `npy_hypotl`:
https://github.com/numpy/numpy/blob/0f7f313d847e8e36e9fb3c4e576619a3af3bfe56/numpy/core/src/npymath/npy_math_internal.h.src#L191-L219
Where is `npy_longdouble npy_hypotl(npy_longdouble, npy_longdouble)` defined if `HAVE_HYPOTL` is not `#define`d?

DWesl committed:
BUG: Set default hypotl to use npy_longdouble arithmetic.

The implementation is already there, and the tests require
npy_longdouble arithmetic, so I set up the boilerplate to make it so.

It seems to fix only np.abs(npy_clongdouble), not
abs(npy_clongdouble), for reasons I don't understand.

DWesl commented:
Found the `npy_hypotl` implementation:
https://github.com/numpy/numpy/blob/bb745921ac6c27d6caf9b76663668e5b9f28abac/numpy/core/src/npymath/npy_math_internal.h.src#L400-L413
It defaults to `npy_hypot`.  Since the tests for `numpy.abs(numpy.clongdouble)` require long double precision and `npy_cabsl` falls back to `npy_hypotl` if the former is not defined, I took the liberty of changing `npy_math_internal.h.src` to repeat the definition of `npy_hypot` with the new types instead:
https://github.com/numpy/numpy/blob/2328404ca5e114e3881184a1e5eee6fba2f739b4/numpy/core/src/npymath/npy_math_internal.h.src#L191-L226
For some reason, this only fixes the `np.abs(np.clongdouble)` tests, not the `abs(np.clongdouble)` ones, at least locally.

DWesl commented:
Should I leave the redefinition of `npy_hypot[fl]` in this PR, or should I move it to a different PR?

I should perhaps mention the two failing checks appear to be the two CircleCI builds that have been failing intermittently over the course of this PR.

DWesl committed:
Go back to old npy_hypotl and mark the failing test case.

DWesl commented:
I went back to the old definition of `npy_hypotl` in terms of `npy_hypot` and marked the `npy_cabsl` tests xfail, since those expect `long double` precision rather than merely double.

DWesl commented:
Is there a preference on whether I change the `cabsl` test to match the current fallback implementation (`npy_hypotl`->`npy_hypot`) or change the `hypotl` fallback implementation to pass the current `cabsl` test?  I currently have the test matching the implementation, but I can go back to having the implementation match the test.  

The CI failures seem unrelated.

charris commented:
The circleci failure should be fixed by a rebase. You might want to merge some of the commits first, though, otherwise you could end up fixing any merge failures in all 38 commits :) Let me try a close/reopen first.

DWesl commented:
The CircleCI failures are still there, so here's one rebase.  I merged many of the commits where I was trying to figure out YAML syntax and how that got passed to PowerShell, but only ran into merge conflicts with other people's code in the last two commits.

Did the "SIMD: Force inlining all functions that accept AVX registers" commit get rolled back?  I think I grabbed that from `main` so the tests would stop segfaulting.

charris commented:
> Should I try reverting that commit and seeing if the tests still pass without segfaults?

You could try. I don't see an indication that it was reverted, I do know that we have been inlining more SIMD functions because some compilers fail to do so, leading to problems.  You could just check out the relevant files from main for such a check,
```
git checkout main <filename>
``` 
which will put them in the index for a commit (IIRC) and also overwrite any local changes when merged. If fixes are needed after that they should probably go into a separate PR.

DWesl committed:
Undo the remaining changes from "SIMD: Force inlining all functions that accept AVX registers"

These changes are not present in `main`.  I see no commits likely to
have specifically changed whether these SIMD functions are inlined.
Adding these back to `main` is left for another PR.  The symptoms I
saw were segfaults, basically because function calls do not preserve
alignment information.

DWesl commented:
> > Should I try reverting that commit and seeing if the tests still pass without segfaults?
> 
> You could try. I don't see an indication that it was reverted, I do know that we have been inlining more SIMD functions because some compilers fail to do so, leading to problems. You could just check out the relevant files from main for such a check,
> 
> ```
> git checkout main <filename>
> ```
> 
> which will put them in the index for a commit (IIRC) and also overwrite any local changes when merged. If fixes are needed after that they should probably go into a separate PR.

I did that, and the tests that had been segfaulting still pass, so I pushed it up for CI.

DWesl commented:
That seems to have removed too many function inline annotations.  The Cygwin test reports a segfault while executing

> File "/usr/lib/python3.7/site-packages/numpy/core/tests/test_api.py", line 20 in test_array_array
> > ```python
> >     ones11 = np.ones((1, 1), np.float64)
> >     assert_equal(np.array(ones11, dtype=np.float64), ones11)
> > ```

I suspect one of the annotations was for an SIMD equality comparison loop.

charris commented:
The next step would be to make a PR adding the annotations that you think are needed, then we can ask @seiko2plus for a review.

charris commented:
It could be that the relevant changes were lost unintentionally.

DWesl commented:
My suspicion is they were lost in merge conflict resolution.  I can cherry-pick the original commit again as a new PR.

DWesl commented:
I added that commit in a PR as #19529.

After digging way back through the pre-rebase commits, I found the original as 05598082761d172f9a5c37298f57cb68ece73386, which seems to have been originally authored on this branch.  There was a different commit that I stole from `main`, which git did indeed drop from the rebase.

DWesl committed:
Revert "Undo the remaining changes from "SIMD: Force inlining all functions that accept AVX registers""

This reverts commit bbd571a0f2b6a81eb69fca6183070cdd4676c5b1.

That commit was original to this branch, but is now in main, so
undoing it is a bad idea.

DWesl commented:
Rebasing to include that commit again seems to have gotten the tests working again.  Anything else I need to do?

One of the changes since the original version of this PR was to add `fmodl` to the list of functions that default to `double` precision if `long double` is not available, so I'm fairly confident leaving `hypotl` in that list and changing the test to match was the right decision.

charris commented on numpy/core/tests/test_scalarmath.py:
```diff
@@ -678,11 +678,35 @@ def _test_abs_func(self, absfunc, test_dtype):
+            for dtype in floating_types + complex_floating_types
+        ],
+    )
     def test_builtin_abs(self, dtype):
```
A cleaner way to do this is to raise skip inside the function. Something like
```
def test_builtin_abs(self, dtype):
    if dtype == np.clongdouble and sys.platform == "cygwin" :
        pytest.skip("absl is computed in double precision on cygwin")
    ...
```
You could also use `pytest.xfail` instead of `pytest.skip`.

DWesl replied:
Done, here and for the next test.

charris commented:
Could move some of the github workflow into a `tools/cygwin_test.sh`, see `.github/actions/action.yml` for an example.  But that can wait.

DWesl committed:
TST: Move the DLL checker to a separate script

Instead of trying to write it every run, just add it to the repository
and use that.  I'm not sure how Windows git handles line endings; I
suspect it changes things to \r\n, which is not what I want.  I'm on
Cygwin, not MSYS, so everything expects \n.

DWesl committed:
TST: Prettify the cabsl test xfail declaration.

Marking this in the parametrize call gets really big.  Moving this
inside the function is much shorter, even if it won't tell me when a
new Cygwin release makes this obsolete.  Since this only needs to wait
for one Cygwin release, I suppose that's not much of a stretch.

DWesl committed:
Wrap long line and change condition order.

I want the `sys.platform == "cygwin"` check before I try to access
`np.clongdouble`, so the check doesn't crash the test on platforms
where np.clongdouble doesn't exist.

charris commented:
In it goes, thanks @DWesl .
